### ISSUE TITLE
Run Prowler using Fargate

### ISSUE THREAD

**Issue by toniblyx** [prowler-cloud/prowler#299]
Based on info gathered here in issue #293 
https://github.com/toniblyx/prowler/issues/293

**Comment by markuman** [comment]
In this case, prowler should provide an official Dockerfile and also maybe upload it to dockerhub.

```Dockerfile
FROM alpine:3.9

RUN apk --update --no-cache add python3 bash curl
RUN pip3 install --upgrade pip
RUN pip install awscli ansi2html boto3
```

The `ansi2html" is only necessary if prowler should generate an html report and put it to S3 for example.  
Otherwise it will only log to stdout and the result is viewable with cloudwatch logs.

**Comment by toniblyx** [comment]
Hi @markuman  there is already a Dockerfile inside `utils` but I have updated that file with your recommendation instead. Please check it out in devel branch https://github.com/toniblyx/prowler/blob/devel/util/Dockerfile I also published it in Docker Hub https://cloud.docker.com/repository/docker/toniblyx/prowler. Could you please test it and see if that's good for Fargate? I'll also test it soon.

**Comment by toniblyx** [comment]
I'll write here a draft high level process to make Prowler run in Fargate as reference. This process has been described and tested by @jevalenciap

_A proper document linked to README.md has to be done as the one for Wazuh._

- Using Dockerfile in master branch: https://github.com/toniblyx/prowler/blob/master/util/Dockerfile 
- Build a image locally and test it. 
- Create an ECR repo and upload the image there.
- Create an IAM role with all needed policies to run Prowler
- Follow the ECS wizard to launch the container based in the pre-built Prowler image. Once it is finished, make sure the proper IAM role is attached to the task definition.
- Once the contaner run Prowler, all results will be in CloudWatch and will be running and keeping logs stored with all the results.

Once that process is done, many other actions can be achieved to find changes between executions, inject it to a DB, etc.

**Comment by chizou** [comment]
Not sure if you'd be interested, but I put together something to run this across all the accounts in our organization using fargate. I have a python script that queries AWS Organizations for a list of accounts then runs this across all the accounts and sends an e-mail via SES to the root e-mail listed. The only problem is it's super slow. It'll usually take over a day to run across all the accounts. I've thought about spawning a bunch of Fargate containers, but I'm worried about API limits since it's all run from one account. 

It also comes with a CodePipeline job to automate deployment new versions of the container/code and CloudFormation templates for deploying everything. 

Let me know if you'd be interested and I'll see about getting it opensourced.

**Comment by toniblyx** [comment]
Of course I'm pretty interested, there is a way to make it faster for multiple accounts, another contributor shared some code to do that and I'm evaluating it now. Also, it depends on how large are your accounts and how many accounts you have but probably an account per container is not bad idea. Can you look at this piece of code written by @jonathanbglass and let me know if it is helpful? I'm planning to make some small changes and include it to the official repo. https://github.com/jonathanbglass/parallel_prowler Thanks!

**Comment by chizou** [comment]
After conferring with colleagues, it doesn't seem rate limiting should be an issue. I'm going to try implementing a multi-container model.

@jonathanbglass's code looks interesting, especially the output parsing. One thing I will note is I originally tried forking multiple processes at once and it killed my container. I didn't use a very big container though so that might have contributed to my issues.

I've asked my boss if we can opensource what we did. Hopefully they give us an approval for that.

**Comment by jonathanbglass** [comment]
I'm thinking about splitting out a combination of accounts times individual
checks, adding them to a queue, then letting lambas launch to do a single
check against a single account. I'd like to see the results get written to
a results queue, and reassembled by a parent process...or write the
individual results to files in s3 and combine them when finished. Then we
can generate the report across multiple accounts.

**Comment by jonrau1** [comment]
I think I have come up with a (mostly) good solution for this. I created a Dockerfile with ENV variables that I pass the AWS CLI Access Key and Secret Access Key to from Secrets Manager. I create plaintext secrets of both, reference the ARNs of the Secrets in ECS Task Definitions and make use of ansi2html to put a colored report in a S3 Bucket.
Feel free to modify as you see fit to close your issue https://github.com/jonrau1/Python-Prowler-Container

**Comment by toniblyx** [comment]
hey @jonrau1 your implementation and documentation is amazing, it would be great if we can add it here as official inside `integrations` folder. What do you think?

**Comment by jonrau1** [comment]
@toniblyx that would be great if you feel it passes muster. I am working on some more documentation behind it, as well as some Terraform scripts to stand up a majority of the solution so some of it may change in the near future

**Comment by toniblyx** [comment]
Despite we will work on a more comprehensive integration this is already done here: https://aws.amazon.com/blogs/security/use-aws-fargate-prowler-send-security-configuration-findings-about-aws-services-security-hub/
